components/brave_private_new_tab_ui/resources/page/hooks.ts

```diff
@@ -0,0 +1,49 @@
+import * as React from 'react'
+import getPageHandlerInstance from './api/brave_page_handler'
+
+export function useTorObserver () {
+  const [hasInit, setHasInit] = React.useState(false)
+  const [isConnected, setIsConnected] = React.useState(false)
+  const [progress, setProgress] = React.useState<string | undefined>(undefined)
+
+  const isLoading = !hasInit && !isConnected
+
+  React.useEffect(() => {
+    getPageHandlerInstance().callbackRouter.onTorCircuitEstablished.addListener(setHasInit)
+    getPageHandlerInstance().callbackRouter.onTorInitializing.addListener(setProgress)
+  }, [])
+
+  React.useEffect(() => {
+    getPageHandlerInstance().pageHandler.getIsTorConnected().then((res: { isConnected: boolean }) => setIsConnected(res.isConnected))
+  }, [hasInit])
+
+  return {
+    isConnected,
+    isLoading,
+    progress
+  }
+}
+
+export function useHasDisclaimerDismissed () {
+  const [hasDisclaimerDismissed, setHasDisclaimerDismissed] = React.useState<boolean>(false)
+
+  React.useEffect(() => {
+    getPageHandlerInstance().pageHandler.getDisclaimerDismissed().then((res: { dismissed: boolean }) => setHasDisclaimerDismissed(res.dismissed))
+  }, [])
+
+  return { hasDisclaimerDismissed }
+}
+
+export function useWaitForMainThreadAndOneFrame () {
+  // We recieve the user pref from C++, which is async and has a ~45ms delay after the main thread executes
+  // Since we want to sync potential UI elements with the decision to show the disclaimer widget, we create a flag, push to event loop and wait for at least one frame
```

Comment:
@petemill I hope this is clear

Reply:
If all we're trying to do is not display components until we've fetched data for the component, then I suggest we just do that. In other words, initialize state properties to `null` (or `undefined`) for API calls which fetch data, and when the data is in that initial state then do not show the relevant UI.